### ISSUE TITLE
`ptr-as-ptr` suggests invalid code for `ptr::null_mut()`

### ISSUE THREAD

**Issue by kyrias** [rust-lang/rust-clippy#11665]
### Summary

The `ptr-as-ptr` lint suggests invalid code for calls to `std::ptr::null_mut()`.

### Reproducer

I tried this code:

```rust
let a = std::ptr::null_mut() as *mut u8;
```

I expected to see this happen:
- The lint should suggest the following code:
  ```rust
  let a = std::ptr::null_mut::<u8>();
  ```

Instead, this happened:

- The lint suggests the following code which doesn't compile:
  ```rust
  let a = std::ptr::null_mut().cast::<u8>();
  ```

### Version

```text
rustc 1.71.0-nightly (4ca000ac8 2023-07-13) (1.71.0.1)
binary: rustc
commit-hash: 4ca000ac85a4f486ed080c281c86ea2287592027
commit-date: 2023-07-13
host: x86_64-unknown-linux-gnu
release: 1.71.0-nightly
LLVM version: 16.0.4
```

**Comment by koka831** [comment]
Presumably it's fixed by https://github.com/rust-lang/rust-clippy/pull/11288, I couldn't reproduce it in `rustc 1.75.0-nightly`.

**Comment by y21** [comment]
I don't think that fixed this. It looks like the PR changed how the type is printed (taking a snippet instead of prettyprint), but the problem here is where the turbofish is placed. The suggested fix `ptr::null_mut().cast::<u8>()` doesn't compile since only the destination pointee type is known. The generic argument should be placed on the `null_mut` call (and just omit `cast()`).
